### PR TITLE
Initial implementation of possible shortcircuiting collect lint

### PR DESCRIPTION
**Pull request by marcusklaas**
This lint detects collect calls that generate collections of `Option` or `Result` and suggests instead collect into a `Option<C>` or `Result<C, _>` respectively, which has personally almost always been what I intend to do. Since this shortcircuiting feature of collect is not very discoverable, I believe this lint has some value.

There may be cases where one really does want a collection of result/ option, in which case this lint will produce a false positive. How should we deal with those? Below is a non-exhaustive list of options:

- make the lint allow by default
- only trigger lint when the iterator item type and its enclosing function type match (either both options or of some type `Result<_, E>`)
- do not trigger lint when it is explicitly stated that we want a collection of options/ results, for example in
```rust
let collection: Vec<Option<_>> = some_iter().collect();
```

Oh, and mega thanks to @eddyb for his gracious guidance through the labyrinths of the rust compiler on the rust IRC!

### CODE REVIEW

File: tests/ui/collect.stderr [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+8 |     let option_vec: Result<std::vec::Vec<_>, _> = b.into_iter()
+  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: you are creating a collection of `option`s
+  --> $DIR/collect.rs:26:18
+   |
+26 |     let _result: Vec<_> = a.iter().map(Some).collect();
+   |                  ^^^^^^
+   |
+help: if you are only interested in the `Some` values or the first `None`, try
```
Comment (oli-obk) [inline tests/ui/collect.stderr]: The message about the first None seems a little weird

File: clippy_lints/src/collect.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+            return if match_type(cx, normal_ty, &paths::OPTION) {
+                Some(Suggestion {
+                    pattern: format_suggestion_pattern(cx, collect_ty.sty.clone(), true),
+                    type_colloquial: "option",
+                    success_variant: "Some",
+                    error_variant: "None",
+                })
+            } else if match_type(cx, normal_ty, &paths::RESULT) {
+                Some(Suggestion {
```
Comment (oli-obk) [inline clippy_lints/src/collect.rs]: Why are these lowercase?

**Comment by oli-obk** [comment]
ping @marcusklaas

**Comment by marcusklaas** [comment]
Hi! Sorry about leaving this for such a long time. Given this lint's inherent propensity of generating false positives, maybe it's better to just close this?

**Comment by oli-obk** [comment]
> Given this lint's inherent propensity of generating false positives, maybe it's better to just close this?

I'd have added it as allow-by-default for now. I like the idea, and maybe we can run it and figure out in which cases we don't want to lint and add them as exceptions

**Comment by marcusklaas** [comment]
Okay. That sounds fair. I will touch up this PR then!

File: clippy_lints/src/collect.rs (hunk 2 above)
Comment (marcusklaas) [reply clippy_lints/src/collect.rs]: Addressed.

File: tests/ui/collect.stderr (hunk 1 above)
Comment (marcusklaas) [reply tests/ui/collect.stderr]: Agreed. Fixed.

**Comment by marcusklaas** [comment]
Build failure seems unrelated to PR. Master branch also fails to build.

**Comment by oli-obk** [comment]
Could you rebase this and use the declare_clippy_lint macro and make this lint a nursery lint?

**Comment by mati865** [comment]
@marcusklaas ping

**Comment by marcusklaas** [comment]
The code appears to be broken and I do not know nearly enough about the compiler internals to fix it.

This PR is best closed I think.

**Comment by oli-obk** [comment]
Does somebody else want to take over? idk what the brokenness is, but it would be a shame to let this PR go to waste! It was pretty close to merging already during it's history, we should be able to get it there again.

**Comment by flip1995** [comment]
Rebased it and fixed a few bugs, which caused ICEs in the `ui/infinite_iter` and the `dogfood` test.

**Comment by flip1995** [comment]
This lint still produces an ICE in rustfmt ([travis](https://travis-ci.org/rust-lang-nursery/rust-clippy/jobs/389682760#L729)).

This error is produced by this line:
https://github.com/marcusklaas/rust-clippy/blob/36993f89598e45ff7007568d0d188e7311c785d8/clippy_lints/src/collect.rs#L121-L124

**Comment by phansch** [comment]
Judging by the ICE message and the documentation of [TyCtxt.normalize_erasing_regions](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.normalize_erasing_regions), I think it's safe to say that the "normalization failed" path was hit. Possibly somewhere in rustc-rayon?

**Comment by phansch** [comment]
Rebased it on master again, let's see if it still fails

File: clippy_lints/src/collect.rs [hunk 3]
```diff
@@ -69,11 +75,11 @@ struct Suggestion {
 
 fn format_suggestion_pattern<'a, 'tcx>(
     cx: &LateContext<'a, 'tcx>,
-    collection_ty: &TypeVariants,
+    collection_ty: &TyKind<'_>,
```
Comment (eddyb) [inline clippy_lints/src/collect.rs]: `TyKind` should never be used directly, you can write `Ty` instead.
@Manishearth @oli-obk I think clippy might have a general problem wrt this.

File: clippy_lints/src/collect.rs [hunk 4]
```diff
@@ -84,7 +90,7 @@ fn format_suggestion_pattern<'a, 'tcx>(
 
             buf
         },
-        TypeVariants::TyParam(p) => p.to_string(),
+        TyKind::Param(p) => p.to_string(),
```
Comment (eddyb) [inline clippy_lints/src/collect.rs]: You should write these as e.g. `ty::Param`.

**Comment by phansch** [comment]
Removed the usage of `TyKind` in https://github.com/rust-lang/rust-clippy/pull/2077/commits/5b7c2064f130dd905d1c0af36b6d47ce33440cba.

The ICE in rustfmt/rls is still there and it's caused by an incorrect use of `normalize_erasing_regions` here:

https://github.com/rust-lang/rust-clippy/blob/4f69870d899817d08500558f998f1bab9d0a798d/clippy_lints/src/collect.rs#L124-L130

I'm not sure what exactly needs to be fixed, but I will try to fix it later today.

**Comment by flip1995** [comment]
ping from triage @phansch: Any updates on this? It seems that this just needs a stderr update.

**Comment by phansch** [comment]
thanks @flip1995, that's news to me. I think I got stuck on fixing the ICE last time but it indeed appears to be gone. I will have another look today.

**Comment by phansch** [comment]
So, currently compiling `rustc-rayon` fails with [the previous ICE](https://github.com/rust-lang/rust-clippy/pull/2077#issuecomment-444382273). The last time I tried, I failed to extract a minimal repro from `rustc-rayon` but I'm going to try it again tomorrow.

**Comment by flip1995** [comment]
half yearly ping @phansch. Should we `S-inactive-closed` this PR?

**Comment by phansch** [comment]
Yeah, makes sense! I currently don't have time to continue this. IIRC, if someone wants to continue this, they should try to find a minimal repro of the ICE that doesn't require `rustc-rayon` and go on fixing the remaining issues from there.

**Pull request closed without merging**